pcsx2/Counters.cpp

```diff
@@ -394,58 +395,65 @@ void frameLimitReset()
 	m_iStart = GetCPUTicks();
 }
 
+// Convenience function to update UI thread and set patches. 
+static __fi void frameLimitUpdateCore()
+{
+	GetCoreThread().VsyncInThread();
+	Cpu->CheckExecutionState();
+}
+
 // Framelimiter - Measures the delta time between calls and stalls until a
 // certain amount of time passes if such time hasn't passed yet.
 // See the GS FrameSkip function for details on why this is here and not in the GS.
 static __fi void frameLimit()
 {
-	// 999 means the user would rather just have framelimiting turned off...
-	if( !EmuConfig.GS.FrameLimitEnable ) return;
-
-	u64 uExpectedEnd	= m_iStart + m_iTicks;
-	u64 iEnd			= GetCPUTicks();
-	s64 sDeltaTime		= iEnd - uExpectedEnd;
+	// Framelimiter off in settings? Framelimiter go brrr.
+	if (!EmuConfig.GS.FrameLimitEnable)
+	{
+		frameLimitUpdateCore();
+		return;
+	}
 
-	// If the framerate drops too low, reset the expected value.  This avoids
-	// excessive amounts of "fast forward" syndrome which would occur if we
-	// tried to catch up too much.
+	u64 uExpectedEnd	= m_iStart + m_iTicks;	// Compute when we would expect this frame to end, assuming everything goes perfectly perfect. 
+	u64 iEnd			= GetCPUTicks();		// The current tick we actually stopped on.
+	s64 sDeltaTime		= iEnd - uExpectedEnd;	// The diff between when we stopped and when we expected to.
 
-	if( sDeltaTime > m_iTicks*8 )
+	// If frame ran too long...
+	if (sDeltaTime >= m_iTicks)
 	{
-		m_iStart = iEnd - m_iTicks;
+		// ... Fudge the next frame start over a bit. Prevents fast forward zoomies.
+		m_iStart += (sDeltaTime / m_iTicks) * m_iTicks;
+		frameLimitUpdateCore();
 		return;
 	}
 
-	// use the expected frame completion time as our starting point.
-	// improves smoothness by making the framelimiter more adaptive to the
-	// imperfect TIMESLICE() wait, and allows it to speed up a wee bit after
-	// slow frames to "catch up."
+	// Conversion of delta from CPU ticks (microseconds) to milliseconds
+	s32 msec = (int) ((sDeltaTime * -1000) / (s64) GetTickFrequency());
+	
+	// If any integer value of milliseconds exists, sleep it off.
+	// Prior comments suggested that 1-2 ms sleeps were inaccurate on some OSes;
+	// further testing suggests instead that this was utter bullshit. 
```

Comment:
This is true and might compromise accuracy a bit - it's worth running a test without sleep at all.

Reply:
I just tested the theory, it does indeed make it smoother, how noticeable it'll be though is any bodies guess

![sleepmsecs](https://user-images.githubusercontent.com/6278726/95016540-7d557b00-064b-11eb-9430-76eeabf37d83.png)

Reply:
Other games such as SotC don't seem to care too much
![sleepmsecs2](https://user-images.githubusercontent.com/6278726/95016869-bee72580-064d-11eb-83ce-3ca522df4089.png)

Reply:
Might be that those games don't hit the condition for a sleep?

Reply:
I run at well above the frame cap in those games. I get about 70fps in SotC, so there should be about 3.334ms my CPU is waiting every frame.

Reply:
Interesting, maybe then you are hitting some non-contractual behaviour of `Sleep` where it special cases smaller sleeps? Even if this was the case I would not rely on it.

Reply:
Well from my experience (and what I've read around the web) the inaccuracy of sleep is < 1ms, I would guess I'm seeing maybe 0.3ms inaccuracy which I don't think anybody is going to honestly notice, usually for stutters you need to see a good couple of ms to notice it.  I think pandubz is going to do this change anyway (maybe some others, dunno) so that will reduce it, I just don't know how much that'll affect people on the edge of the frame cap, spinlocking is great but it uses CPU power to do it.

Reply:
It might depend on the power plan as my experience with sleeping inside a frame limiter was drastically different - to the extent where e.g. a loop intended to lock to 120 FPS ended up at 112 FPS.

Reply:
yeah I mean that's 0.59ms which works out as 2fps either way, which isn't brilliant I'd admit, but as experienced, I changed it to do if(msec > 1) sleep(msec-1) so there was always at least 1ms we had to spinlock on, that got me the above results, which didn't improve things drastically.

You're not going to run PS2 games at 120fps that's for sure :P

Reply:
Another look at the Xenosaga Ep. 1 Elsa save point. Master (top), `msec >=1 then sleep(msec)` (middle), `msec > 1 then sleep(msec - 1)` (bottom). I think we have our winner.
![image](https://user-images.githubusercontent.com/6377490/95020105-16c66200-0637-11eb-9190-bc0379747c74.png)

Reply:
That's a good approach, but I would go even further and make it something like `msec - 3` or even 5. The threshold has to be pretty big or else you'll always risk people losing time on those e.g. due to running on a laptop battery.

Reply:
Well I don't recommend people run the emulator on a laptop and if they do they should be using high performance mode and not running off battery (that's a really dumb idea :P) I'm honestly not concerned about people crazy enough to do that, it will still be head over shoulders above master (which was 1ms buffer)

Reply:
Sure, that should be good enough as a starting point and could be revisited in the future if it's ever needed.

Reply:
I'm using a gaming laptop plugged in with high performance mode and it does improve significantly on specific games, u could stare blindly at squeezing an extra 5%-10%.

Reply:
Good timing, as Bruce Dawson just published a somewhat-related post:
https://randomascii.wordpress.com/2020/10/04/windows-timer-resolution-the-great-rule-change/